src/v/cloud_storage/remote_partition.cc

```diff
@@ -133,9 +134,14 @@ class partition_record_batch_reader_impl final
                     vlog(_ctxlog.debug, "Unexpected error");
                     throw std::system_error(result.error());
                 }
-                // empty result will also be propagated here
-                data_t d = std::move(result.value());
-                co_return storage_t{std::move(d)};
+                if (result.value().empty()) {
+                    _partition->evict_reader(std::move(_reader));
```

Comment:
another option is to call `co_await set_end_of_stream()`
it will guarantee that subsequent call will return immediately

Reply:
Fixed.